core/src/main/java/com/digitalpebble/stormcrawler/parse/TextExtractor.java

```diff
@@ -138,14 +143,71 @@ public void tail(Node node, int depth) {
                         }
                     }
                 },
-                node);
-        return accum.toString().trim();
+                node,
+                maxTextSize,
+                accum);
     }
 
-    private static void appendNormalisedText(StringBuilder accum, TextNode textNode) {
-        String text = textNode.getWholeText();
+    /**
+     * Start a depth-first traverse of the root and all of its descendants.
+     *
+     * @param visitor Node visitor.
+     * @param root the root node point to traverse.
+     */
+    public static void traverse(
+            NodeVisitor visitor, Node root, int maxSize, StringBuilder builder) {
+        Validate.notNull(visitor);
+        Validate.notNull(root);
+        Node node = root;
+        int depth = 0;
+
+        while (node != null) {
+            // interrupts if too much text has already been produced
+            if (maxSize > 0 && builder.length() >= maxSize) return;
+
+            Node parent =
+                    node.parentNode(); // remember parent to find nodes that get replaced in .head
+            int origSize = parent != null ? parent.childNodeSize() : 0;
+            Node next = node.nextSibling();
+
+            visitor.head(node, depth); // visit current node
+            if (parent != null && !node.hasParent()) { // removed or replaced
+                if (origSize == parent.childNodeSize()) { // replaced
+                    node =
+                            parent.childNode(
+                                    node.siblingIndex()); // replace ditches parent but keeps
+                    // sibling index
+                } else { // removed
+                    node = next;
+                    if (node == null) { // last one, go up
+                        node = parent;
+                        depth--;
+                    }
+                    continue; // don't tail removed
+                }
+            }
+
+            if (node.childNodeSize() > 0) { // descend
+                node = node.childNode(0);
+                depth++;
+            } else {
+                while (true) {
+                    assert node != null; // as depth > 0, will have parent
+                    if (!(node.nextSibling() == null && depth > 0)) break;
```

Comment:
IMHO it would be more readable to bring this breaking condition into the loop instead of using `true` there.